File: nightly/tests/c/branchflow.c
/* Taint that reaches the sink on only one path, the C analogue of
 * tests/java/BranchingFlow.java. `result` is tainted through the then-branch
 * and clean through the else-branch; the sink is reachable either way, so the
 * flow must still be reported (the analysis is not path sensitive).
 *
 * Only the source and the sink are in expected_lines: the tainted branch's
 * assignment is on the flow, but at -O0 no reported instruction maps back to
 * it.
 */

int source(void);
void sink(int x);

int main(int argc, char **argv) {
  (void)argv;
  int data = source();
  int result;
  if (argc > 1) {
    result = data;
  } else {
    result = 0;
  }
  sink(result);
  return 0;
}

File: nightly/tests/lua/branch-flow.lua
-- Control flow: the tainted value is assigned on one branch of an if/else. The
-- join after the branch carries the taint to the sink.

local function source()
  return io.read()
end

local function sink(x)
  print(x)
end

local function main(flag)
  local value
  if flag then
    value = source()
  else
    value = "default"
  end
  sink(value)
end

main(true)
